### PR TITLE
polygon support

### PR DESCRIPTION
**Pull request by will-moore**
Supports adding of Polygons from OMERO to figure and editing them in Figure.

NB: creation of NEW polygons in figure is not supported in this PR.

To test:
 - Create and save Polygons AND Polylines in:
     - iviewer
     - Insight
     - ImageJ.
 - Open image in Figure
 - Add ROIs -> Load from OMERO -> Pick Polygon / Polyline
 - Once added, can edit (drag shape or corner handles)
 - Test Copy and Paste of Polygons/Polylines, between panels AND within 1 panel
 - Save figure, reload etc - check Polygons/Polylines saved OK.
 - Export to TIFF and PDF   *NB: Polyline not supported here yet

cc @pwalczysko 

screen shot 2017-11-20 at 09 51 06

### CODE REVIEW

**Comment by jburel** [comment]
let's put that onhold, we will do without for training.
We need to discuss the future of our own drawing library
ROI handling in figure in general needs to be reviewed.

**Comment by jburel** [comment]
To add to the description:
polygon should be created from different source: minimally insight and imageJ
since I noticed some difference between applications see https://github.com/openmicroscopy/openmicroscopy/pull/5597

**Comment by manics** [comment]
Don't forget there's the really old format too: https://github.com/openmicroscopy/openmicroscopy/blob/v5.4.1/components/tools/OmeroPy/test/unit/test_rois.py#L32 which should be either tested or an appropriate error message shown.

**Comment by will-moore** [comment]
Thanks @manics - That should be handled by the webgateway marshalling: https://github.com/openmicroscopy/openmicroscopy/blob/develop/components/tools/OmeroWeb/omeroweb/webgateway/marshal.py#L357

**Comment by jburel** [comment]
The same should be done for Polyline

**Comment by jburel** [comment]
The changes in shape-editor must go to its repository https://github.com/ome/shape-editor
A new version will be released prior to adding the feature in figure

File: Gruntfile.js [hunk 1]
```diff
@@ -138,7 +138,19 @@ module.exports = function (grunt) {
...
+          },
+        ]
+      },
+      shapeEditor: {
+        files: [
+          {
+            expand: true,
+            cwd: 'node_modules/ome-shape-editor/dist/js',
+            src: 'shape-editor.js',
+            dest: 'omero_figure/static/figure/3rdparty/shape-editor-3.1.0'
```
Comment (jburel) [inline Gruntfile.js]: The version should ideally be retrieved from package.json

Comment (will-moore) [reply Gruntfile.js]: Hmm - that might be kinda tricky.
This version number is also in index.html
```
<script src="{% static 'figure/3rdparty/shape-editor-3.1.0/shape-editor.js' %}"></script>
```
Having that version number in multiple places is not new, but now we don't have the shape-editor.js code itself.

**Comment by will-moore** [comment]
@jburel Any idea what's going on here? Even when I remove the small change I made to ```setup.py```, travis is still failing here with
```
+ python setup.py sdist
running sdist
running grunt
running npm_install
npm install
unable to execute npm: No such file or directory
error: command 'npm' failed with exit status 1
```

**Comment by jburel** [comment]
I will need to check the changes.

**Comment by jburel** [comment]
@will-moore it no longer works when we combine app and scripts
i.e. ``  - .omero/docker app`` works fine
The work done in https://github.com/ome/omero-figure/pull/270 and @joshmoore test-infra branch should allow to have both working again.
you can temporarily update the travis file

**Comment by pwalczysko** [comment]
This is at the moment not testable on eel-merge. OMERO.figure on eel-merge is somehow not right:

   - In webclient go Figure. In Figure, go File > Open...  No new open dialog pops up. 
   - in webclient select an image and in context menu select Open With OMERO.figure
   - a new tab opens as expected, but the Figure does not

**Comment by will-moore** [comment]
This is because ```shape-editor.js``` is 404. Obviously something is going wrong with change in build step on this PR.

**Comment by pwalczysko** [comment]
Interestingly, OMERO.figure can upload only 200 ROIs.

**Comment by pwalczysko** [comment]
Polygon export problem:

   - export pdf or tiff of http://web-dev-merge.openmicroscopy.org/figure/file/34053 figure (username: outreach)
   - observe distortions of the polygons in the exported (long straight line "jumps" pointing towards the botoom left of the image)

screen shot 2018-03-08 at 11 03 20

**Comment by pwalczysko** [comment]
The Copy > Paste workflow:


   - when a polyline is selected and copied, the Paste button which is not in the ``Edit ROIs`` window, but in the ``Labels`` tab itself, often results in ``Paste failure`` (see screenshot below)
   - when the same Paste is later attempted using the Edti > Paste Shape workflow inside the ``Edit ROIs`` window, the paste succeeds
   - see http://web-dev-merge.openmicroscopy.org/figure/file/34055,  user; outreach, try to paste the left-most polyline in the top image only

screen shot 2018-03-08 at 12 11 24

**Comment by pwalczysko** [comment]
2 problems detected https://github.com/ome/omero-figure/pull/253#issuecomment-371455598 and https://github.com/ome/omero-figure/pull/253#issuecomment-371470339, otherwise works as expected

**Comment by rgozim** [comment]
Tested locally and working is working fine.

**Comment by joshmoore** [comment]
Raising the limit to 500 seems fairly arbitrary and likely to require your testers to increase the number of ROIs they have to add before they see a potential problem, no?

**Comment by will-moore** [comment]
Not entirely arbitrary, but this was a quick fix I needed for the [iviewer demo movie](https://trello.com/c/kud8kpX4/28-movie-demo) and could certainly be improved.
500 is the default max limit of the JSON API allows, so setting it higher than that here would not allow you to load any more ROIs.
I could get the max limit from ```api``` and use that, so that if someone were to config a higher value then it would allow more to be loaded here too.
Unfortunately there is only 1 ```API_MAX_LIMIT``` setting which applies to all objects, so it's not possible to allow loading of 1000 ROIs while keeping the limit lower for e.g. Wells or Images (which tend to be more verbose).

**Comment by joshmoore** [comment]
> was a quick fix I needed for the iviewer demo movie

understood.

> 500 is the default max limit of the JSON API allows, so setting it higher than that here would not allow you to load any more ROIs.

understood. Was having it at 200 causing issues because it was _different_ from the API's 500 or is there another underlying issue? i.e. must the values be the same?

**Comment by jburel** [comment]
I reckon 200 was an arbitrary value set since figure did not use the json api to load roi.

**Comment by will-moore** [comment]
On the image I used in the movie, the ImageJ Analyse-Particles created over 200 polygons on the image, so that the Arrow I added in iviewer was not included in the first 200 shapes I loaded into figure (so it didn't show up in figure).
200 is the default ```API_LIMIT``` of the json api.

**Comment by joshmoore** [comment]
Sorry for derailing this PR. Assuming the summary is correct, happy for "figure should load more than 1 page of ROIs" to be captured somewhere.

**Comment by will-moore** [comment]
That should fix the 2 problems reported by @pwalczysko above.

File: omero_figure/scripts/omero/figure_scripts/Figure_To_Pdf.py [hunk 2]
```diff
@@ -148,29 +152,29 @@ def panel_to_page_coords(self, shape_x, shape_y):
...
+            # img coords: centre of rotation
+            cx = self.crop['x'] + (self.crop['width']/2)
+            cy = self.crop['y'] + (self.crop['height']/2)
+            dx = cx - shape_x
+            dy = cy - shape_y
+            # distance of point from centre of rotation
+            h = sqrt(dx * dx + dy * dy)
+            # and the angle (avoid division by zero!)
+            if dy == 0:
+                dy = 0.000001
```
Comment (jburel) [inline omero_figure/scripts/omero/figure_scripts/Figure_To_Pdf.py]: what is the rationale?

Comment (will-moore) [reply omero_figure/scripts/omero/figure_scripts/Figure_To_Pdf.py]: You can't divide by zero, but dividing by a number close to zero works very well and is the simplest workaround (the previous workaround was not working).

Comment (sbesson) [reply omero_figure/scripts/omero/figure_scripts/Figure_To_Pdf.py]: Isn't this the case we you want to use `math.atan2` rather than `math.atan`

Comment (will-moore) [reply omero_figure/scripts/omero/figure_scripts/Figure_To_Pdf.py]: Yep, that works. Thanks

**Comment by pwalczysko** [comment]
>That should fix the 2 problems reported by @pwalczysko above.

Yes, both problems are fixed. Ready to merge FMPOV

**Comment by jburel** [comment]
Could you open against shape-editor?
We first need to tag shape-editor before moving forward with this PR

**Comment by snoopycrimecop** [comment]
Conflicting PR. Removed from build [FIGURE-merge#677](https://ci.openmicroscopy.org/job/FIGURE-merge/677/). See the [console output](https://ci.openmicroscopy.org/job/FIGURE-merge/677/consoleText) for more details.
Possible conflicts:
  - PR #243 will-moore 'Big images'
    - omero_figure/templates/figure/index.html

--conflicts

**Comment by will-moore** [comment]
Doh - conflict was bound to happen sometime I guess with these two big PRs.
But shape-editor PR is open in the meantime...

**Comment by jburel** [comment]
No problem noticed with handling of polygons/polylines
next steps are:
 * clean up code in https://github.com/ome/shape-editor
 * tag and push a 3.1.1 version to npm using the ``openmicroscopy`` scope https://github.com/ome/shape-editor/pull/9
* update the version of shape-editor used in this PR

**Comment by will-moore** [comment]
I have deprecated shape-editor from npm (couldn't unpublish).
```
$ npm deprecate -f 'ome-shape-editor@3.1.0' "To be replaced by openmicroscopy/shape-editor"
```
Can you now publish ```openmicroscopy/shape-editor``` so I can use that?

**Comment by jburel** [comment]
package has now been published under the organisation
https://www.npmjs.com/settings/openmicroscopy/packages

**Comment by will-moore** [comment]
Thanks. Hopefully with a final test to check this is working, we should be good to merge?

**Comment by jburel** [comment]
I would think so. Added chat to standup

**Comment by jburel** [comment]
Retested with the latest change on web-dev-merge
Still not great to have the version in 3 different places with it
Merging so we can move forward with the Big images review

**Pull request merged**